akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/R2dbcProjectionImpl.scala

```diff
@@ -667,61 +669,66 @@ private[projection] object R2dbcProjectionImpl {
-                      else FutureNotReplayed
+                        SourceReplayTriggered
+                      else SourceNotReplayed
                   }
                 }
-
               case _ =>
                 triggerReplayIfPossible(sourceProvider, offsetStore, originalEnvelope).map {
-                  case true  => ReplayTriggered
-                  case false => NotReplayed
+                  case true  => SourceReplayTriggered
+                  case false => SourceNotReplayed
                 }(ExecutionContext.parasitic)
             }
+            fastFutureSource(futureSource)
           case _ =>
-            FutureNotReplayed // no replay support for non typed envelopes
+            SourceNotReplayed // no replay support for non typed envelopes
         }
+
       } else {
         // replayOnRejectedSequenceNumbers is disabled
-        FutureNotReplayed
+        SourceNotReplayed
       }
     }
 
     def observer(context: ProjectionContext): HandlerObserver[Envelope] =
       context.asInstanceOf[ProjectionContextImpl[Offset, Envelope]].observer
 
-    def accepted(envelope: Envelope, context: ProjectionContext): Future[Option[(Envelope, ProjectionContext)]] = {
+    def accepted(
+        envelope: Envelope,
+        context: ProjectionContext): Source[Option[(Envelope, ProjectionContext)], NotUsed] = {
       if (isFilteredEvent(envelope) && settings.warnAboutFilteredEventsInFlow) {
         log.info("atLeastOnceFlow doesn't support skipping envelopes. Envelope [{}] still emitted.", envelope)
       }
-      loadEnvelope(envelope, sourceProvider).map { loadedEnvelope =>
+      fastSourceFuture(loadEnvelope(envelope, sourceProvider)).map { loadedEnvelope =>
         offsetStore.addInflight(loadedEnvelope)
         Some((loadedEnvelope, context))
       }
     }
 
+    val singleNoneSource: Source[Option[(Envelope, ProjectionContext)], NotUsed] = Source.single(None)
     val validate = Flow[(Envelope, ProjectionContext)]
-      .mapAsync(1) {
+      .flatMapConcat {
         case (env, context) =>
+          // Note: several calls here return Sources while actually only ever returning a single value to fold it into
+          // the stream rather than through a future+mapAsync
           offsetStore
-            .validate(env)
-            .flatMap {
-              case Accepted =>
-                accepted(env, context)
+            .validateSource(env)
+            .flatMapConcat {
+              case Accepted => accepted(env, context)
               case Duplicate =>
-                Future.successful(None)
+                singleNoneSource
               case RejectedSeqNr =>
-                replayIfPossible(env, observer(context)).flatMap {
+                replayIfPossible(env, observer(context)).flatMapConcat {
                   // if missing events were replayed immediately, then accept the rejected envelope for downstream processing
                   case ReplayedOnRejection => accepted(env, context)
```

Comment:
ok, looking closer I think I understand. there will only be element from replayIfPossible, and then we append the original envelope. Meaning that offset will not be saved for the intermediate replayed events, but it will be saved for the original. That's same as we had before.

What happens if replayIfPossible is an empty source, which I think can happen if all are duplicates
```
  case Duplicate =>
    None
```

That seems like it shouldn't happen, but would be mysterious to silently drop the original env for that case?